File: cli/asset_detection.py
"""Asset class detection utilities for the CLI."""

# Known commodities from Alpha Vantage
KNOWN_COMMODITIES = {
    "WTI",
    "BRENT",
    "NATURAL_GAS",
    "COPPER",
    "ALUMINUM",
    "WHEAT",
    "CORN",
    "SUGAR",
    "COTTON",
    "COFFEE",
}

# Known cryptocurrencies supported by Alpha Vantage
KNOWN_CRYPTOS = {
    "BTC",
    "ETH",
    "BNB",
    "XRP",
    "ADA",
    "SOL",
    "DOGE",
    "DOT",
    "MATIC",
    "AVAX",
    "LINK",
    "UNI",
    "ATOM",
    "LTC",
    "BCH",
    "XLM",
    "ALGO",
    "VET",
    "ICP",
    "FIL",
}


def detect_asset_class(symbol: str) -> str:
    """
    Automatically detect if a symbol is a commodity, crypto, or equity.

    Args:
        symbol: The ticker symbol (e.g., "BRENT", "BTC", "BTC-USD", "AAPL")

    Returns:
        "commodity", "crypto", or "equity" based on the symbol
    """
    symbol_upper = symbol.upper()

    # Check exact match first
    if symbol_upper in KNOWN_COMMODITIES:
        return "commodity"
    if symbol_upper in KNOWN_CRYPTOS:
        return "crypto"

    # Check if it's crypto with suffix (e.g., BTC-USD, ETH-USD)
    # Remove common crypto suffixes
    for suffix in ["-USD", "-USDT", "-EUR", "-GBP"]:
        if symbol_upper.endswith(suffix):
            base_symbol = symbol_upper[: -len(suffix)]
            if base_symbol in KNOWN_CRYPTOS:
                return "crypto"

    return "equity"


def get_asset_class_display_name(asset_class: str) -> str:
    """Get a human-friendly display name for the asset class."""
    return asset_class.capitalize()


def normalize_ticker(symbol: str, asset_class: str | None = None) -> str:
    """
    Normalize ticker symbol for the appropriate data vendor.

    For crypto: Ensures -USD suffix for yfinance compatibility
    For stocks/commodities: Returns as-is

    Args:
        symbol: Ticker symbol (BTC, BTC-USD, AAPL, etc.)
        asset_class: Asset class (auto-detected if None)

    Returns:
        Normalized ticker symbol
    """
    if asset_class is None:
        asset_class = detect_asset_class(symbol)

    if asset_class == "crypto":
        # Add -USD suffix if not present
        if not symbol.upper().endswith(("-USD", "-USDT", "-EUR", "-GBP")):
            return f"{symbol}-USD"

    return symbol

File: frontend/src/utils/assetDetection.ts
/**
 * Utility functions for detecting asset types from ticker symbols
 */

const CRYPTO_SYMBOLS = new Set([
  'BTC', 'ETH', 'BNB', 'XRP', 'ADA', 'SOL', 'DOGE', 'DOT', 'MATIC', 'AVAX',
  'LINK', 'UNI', 'ATOM', 'LTC', 'BCH', 'XLM', 'ALGO', 'VET', 'ICP', 'FIL',
  'TRX', 'ETC', 'XMR', 'NEAR', 'APT', 'ARB', 'OP', 'SHIB', 'PEPE', 'WIF',
])

const COMMODITY_SYMBOLS = new Set([
  'BRENT', 'WTI', 'GOLD', 'SILVER', 'COPPER', 'PLATINUM', 'PALLADIUM',
  'NATURAL_GAS', 'WHEAT', 'CORN', 'SOYBEANS', 'COFFEE', 'SUGAR', 'COTTON',
])

/**
 * Detect if a ticker is a cryptocurrency
 */
export function isCrypto(ticker: string): boolean {
  const normalized = ticker.toUpperCase().replace(/-USD$/, '')
  return CRYPTO_SYMBOLS.has(normalized)
}

/**
 * Detect if a ticker is a commodity
 */
export function isCommodity(ticker: string): boolean {
  const normalized = ticker.toUpperCase()
  return COMMODITY_SYMBOLS.has(normalized)
}

/**
 * Get the asset class for a ticker
 */
export function getAssetClass(ticker: string): 'crypto' | 'commodity' | 'equity' {
  if (isCrypto(ticker)) return 'crypto'
  if (isCommodity(ticker)) return 'commodity'
  return 'equity'
}

/**
 * Format volume based on asset class
 * - Crypto: Show token units with USD equivalent
 * - Equity: Show share count
 * - Commodity: Hide (not applicable)
 */
export function formatVolume(
  volume: number | null,
  assetClass: 'crypto' | 'commodity' | 'equity',
  ticker: string,
  price?: number
): string | null {
  if (volume === null || volume <= 0) return null

  const formatCompactNumber = (value: number): string => {
    if (value >= 1_000_000_000) return `${(value / 1_000_000_000).toFixed(1)}B`
    if (value >= 1_000_000) return `${(value / 1_000_000).toFixed(1)}M`
    if (value >= 1_000) return `${(value / 1_000).toFixed(1)}K`
    return `${Math.floor(value)}`
  }

  switch (assetClass) {
    case 'crypto': {
      // For crypto, volume is in token units from a single exchange
      // Show both token count and USD value if price is available
      const tokenSymbol = ticker.toUpperCase().replace(/-USD$/, '')
      const tokenVolume = formatCompactNumber(volume)

      if (price && price > 0) {
        const usdVolume = volume * price
        const usdFormatted = formatCompactNumber(usdVolume)
        return `${tokenVolume} ${tokenSymbol} (~$${usdFormatted})`
      }

      return `${tokenVolume} ${tokenSymbol}`
    }

    case 'equity': {
      // For stocks, volume is share count
      return formatCompactNumber(volume)
    }

    case 'commodity': {
      // Commodities don't have meaningful volume data
      return null
    }

    default:
      return formatCompactNumber(volume)
  }
}

